---
title: "sc_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}

library(dplyr)
library(ggplot2)
library(Matrix)
library(patchwork)
library(Seurat)

```



```{r}

# import data
dmso.data <- Read10X(data.dir = "/Users/David/Documents/NYU/ASI/finalproject/output/cellranger_count/dmso/filtered_feature_bc_matrix")
ly.data <- Read10X(data.dir = "/Users/David/Documents/NYU/ASI/finalproject/output/cellranger_count/LY/filtered_feature_bc_matrix")
mirin.data <- Read10X(data.dir = "/Users/David/Documents/NYU/ASI/finalproject/output/cellranger_count/Mirin/filtered_feature_bc_matrix")


```


```{r}

# create Seurat objects from data imports
dmso <- CreateSeuratObject(counts = dmso.data,
                         min.cells=3,
                         min.features = 200,
                         project = "dmso")

dmso@meta.data[, "protocol"] = "dmso"


ly = CreateSeuratObject(counts = ly.data,
                        min.cells=3,
                        min.features=200,
                        project = "ly")

ly@meta.data[, "protocol"] = "ly"


mirin = CreateSeuratObject(counts = mirin.data,
                        min.cells=3,
                        min.features=200,
                        project = "mirin")

mirin@meta.data[, "protocol"] = "mirin"

```

```{r}
# Normalize and find variable features
dmso <- NormalizeData(dmso, verbose=FALSE)
ly <- NormalizeData(ly, verbose = FALSE)
mirin <- NormalizeData(mirin, verbose=FALSE)

dmso <- FindVariableFeatures(dmso, selection.method = "vst", nfeatures = 2000)
ly <- FindVariableFeatures(ly, selection.method = "vst", nfeatures = 2000)
mirin <- FindVariableFeatures(mirin, selection.method = "vst", nfeatures = 2000)



```

```{r}
merged.anchors <- FindIntegrationAnchors(object.list = list(dmso, ly, mirin), dims = 1:20)
merged.combined <-  IntegrateData(anchorset = merged.anchors, dims = 1:20)



```

```{r}
DefaultAssay(merged.combined) <- "integrated"

merged.combined <- ScaleData(merged.combined, verbose = FALSE)
merged.combined <- RunPCA(merged.combined, features = VariableFeatures(object = merged.combined), npcs = 30, verbose = FALSE)

ElbowPlot(merged.combined)

```



```{r}
# Create UMAP and TSNE projections, run clustering

merged.combined <- RunUMAP(merged.combined, reduction = "pca", dims = 1:15)
merged.combined <- RunTSNE(merged.combined, reduction = "pca")
merged.combined <- FindNeighbors(merged.combined, dims = 1:15)
merged.combined <- FindClusters(merged.combined, resolution = 0.08)

```






```{r}
# plots
p1 <- DimPlot(merged.combined, reduction = "tsne", group.by="protocol")
p2 <- DimPlot(merged.combined, reduction = "tsne", label = TRUE)
plot_grid(p1, p2)

DimPlot(merged.combined, reduction = "umap", split.by = "protocol")

```








```{r}
# heatmap of cluster gene expressiiion
DefaultAssay(merged.combined) = "integrated"

merged.markers <- FindAllMarkers(merged.combined,
                                 only.pos = TRUE,
                                 min.pct = 0.25, 
                                 logfc.thrshold = 1)



merged.markers %>% group_by(cluster) %>% top_n(n=2, wt = avg_logFC) 

top10 <- merged.markers %>% group_by(cluster) %>% top_n(n=10, wt = avg_logFC)
DoHeatmap(merged.combined, features = top10$gene) + 
  NoLegend() + 
  ggtitle("Heatmap of Gene Expression, Seurat Automatic Clustering") + 
  theme(axis.text.y = element_text(color = "grey20", size = 4, angle = 0, hjust = 1, vjust = 0, face = "plain"))


```




```{r}
# section for indicating examples of cell-type identification based on abcam markers

abcam_neuroepithelial = c("Nes", "Sox2", "Notch1", "Hes1", "Sox10", "Fabp7", "Mog")

abcam_radialglial = c("Pax6", "Hes1", "Fabp7" ,"Tnc")

abcam_schwann = c("Sox10", "Egr2", "Mpb", "Mpz")

abcam_microglia = c("Ptprc", "Aif1", "Cd68", "Lyz2")

FeaturePlot(merged.combined, features = abcam_microglia)

```



```{r}
# reassign labels based on transcription data
# adapted from  https://rpubs.com/mathetal/integratedanalysis
numbered_copy <- merged.combined
merged.combined.celltypes <- merged.combined


number.cluster.ids = c(0,1,2,3,4,5,6,7,8,9)
celltype.cluster.ids = c("Non-diff", "Neurons", "Neurons", "Radial_Glial_Cells", "Neurons", "Neurons", "Neurons", "Microglia", "NE_Schwann_Cells", "Unknown")

names(x = celltype.cluster.ids) <- levels(x = merged.combined.celltypes)
merged.combined.celltypes <- RenameIdents(object = merged.combined.celltypes, celltype.cluster.ids)

DimPlot(merged.combined.celltypes, reduction = "tsne", label=T, repel=T) + 
  ggtitle("t-SNE Plot of Embryonic Superior Cervical Ganglion \n Cell Population Gene Expression with Cell Type Labeling")

```

```{r}
# detect conserved markers on Seurat automatic clustering
cluster0.markers <- FindConservedMarkers(merged.combined, ident.1 = 0, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster1.markers <- FindConservedMarkers(merged.combined, ident.1 = 1, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster2.markers <- FindConservedMarkers(merged.combined, ident.1 = 2, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster3.markers <- FindConservedMarkers(merged.combined, ident.1 = 3, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster4.markers <- FindConservedMarkers(merged.combined, ident.1 = 4, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster5.markers <- FindConservedMarkers(merged.combined, ident.1 = 5, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster6.markers <- FindConservedMarkers(merged.combined, ident.1 = 6, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster7.markers <- FindConservedMarkers(merged.combined, ident.1 = 7, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster8.markers <- FindConservedMarkers(merged.combined, ident.1 = 8, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
cluster9.markers <- FindConservedMarkers(merged.combined, ident.1 = 9, grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)

```



```{r}

# find markers per annotated cell type
neuron.markers <- FindConservedMarkers(merged.combined.celltypes, ident.1 = "Neurons", grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
microglia.markers <- FindConservedMarkers(merged.combined.celltypes, ident.1 = "Microglia", grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
ne_schwann.markers <- FindConservedMarkers(merged.combined.celltypes, ident.1 = "NE_Schwann_Cells", grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
radialglia.markers <- FindConservedMarkers(merged.combined.celltypes, ident.1 = "Radial_Glial_Cells", grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)
unknown.markers <- FindConservedMarkers(merged.combined.celltypes, ident.1 = "Unknown", grouping = "protocol", logfc.threshold = 1, min_pct = 0.25)

```





```{r}
# expression per cell type
microglia.de <- subset(merged.combined.celltypes, idents="Microglia")
Idents(microglia.de) <- "protocol"
avg.microglia <- log1p(AverageExpression(microglia.de, verbose=FALSE)$RNA)
avg.microglia$gene <- rownames(avg.microglia)

neurons.de <- subset(merged.combined.celltypes, idents="Neurons")
Idents(neurons.de) <- "protocol"
avg.neurons <- log1p(AverageExpression(neurons.de, verbose=FALSE)$RNA)
avg.neurons$gene <- rownames(avg.neurons)

neschwann.de <- subset(merged.combined.celltypes, idents="NE_Schwann_Cells")
Idents(neschwann.de) <- "protocol"
avg.neschwann <- log1p(AverageExpression(neschwann.de, verbose=FALSE)$RNA)
avg.neschwann$gene <- rownames(avg.neschwann)

nonspecific.de <- subset(merged.combined.celltypes, idents="Non-diff")
Idents(nonspecific.de) <- "protocol"
avg.nonspecific <- log1p(AverageExpression(nonspecific.de, verbose=FALSE)$RNA)
avg.nonspecific$gene <- rownames(avg.nonspecific)


```



```{r}
# Create differential expression arrays between test conditions per identified cell type
merged.combined.test <-  merged.combined.celltypes

merged.combined.test$celltype.protocol <- paste(Idents(merged.combined.test), merged.combined.test$protocol, sep = "_")
merged.combined.test$celltype <- Idents(merged.combined.test)
Idents(merged.combined.test) <- "celltype.protocol"


mirin.neuron.response <- FindMarkers(merged.combined.test, ident.1 = "Neurons_dmso", ident.2 = "Neurons_mirin", verbose = FALSE)
mirin.microglia.response <- FindMarkers(merged.combined.test, ident.1 = "Microglia_dmso", ident.2 = "Microglia_mirin", verbose = FALSE)
mirin.schwann.response <- FindMarkers(merged.combined.test, ident.1 = "NE_Schwann_Cells_dmso", ident.2 = "NE_Schwann_Cells_mirin", verbose = FALSE)
mirin.nonspecific.response <- FindMarkers(merged.combined.test, ident.1 = "Non-diff_dmso", ident.2 = "Non-diff_mirin", verbose = FALSE)


ly.neuron.response <- FindMarkers(merged.combined.test, ident.1 = "Neurons_dmso", ident.2 = "Neurons_ly", verbose = FALSE)
ly.microglia.response <- FindMarkers(merged.combined.test, ident.1 = "Microglia_dmso", ident.2 = "Microglia_ly", verbose = FALSE)
ly.schwann.response <- FindMarkers(merged.combined.test, ident.1 = "NE_Schwann_Cells_dmso", ident.2 = "NE_Schwann_Cells_ly", verbose = FALSE)
ly.nonspecific.response <- FindMarkers(merged.combined.test, ident.1 = "Non-diff_dmso", ident.2 = "Non-diff_ly", verbose = FALSE)

```



```{r}

freq_table <- prop.table(x = table(merged.combined.celltypes@active.ident, merged.combined.celltypes@meta.data[, "protocol"]),
                         margin = 2)

write.csv(freq_table, "/Users/David/Desktop/freq_table.csv")

```



